### PR TITLE
Extmaps

### PR DESCRIPTION
**Pull request by jbrady42**
Basic pass-through ext map support.

TODO
- [x] Handle global exts
- [x] SDP Tests
- [x] Const for `SDPSectionType`

### CODE REVIEW

**Comment by Sean-Der** [comment]
nice! So I don't think you need to worry about directions in this PR, extmap doesn't care about.

File: settingengine.go [hunk 1]
```diff
@@ -239,3 +241,14 @@ func (e *SettingEngine) DisableSRTPReplayProtection(isDisabled bool) {
 func (e *SettingEngine) DisableSRTCPReplayProtection(isDisabled bool) {
 	e.disableSRTCPReplayProtection = isDisabled
 }
+
+// AddSDPExtensions adds available and offered extensions for media type.
```
Comment (Sean-Der) [inline settingengine.go:245]: :+1: LGTM!

Comment (Sean-Der) [reply settingengine.go:245]: I would make sure the comment mentions
* Don't include the ID, we generate those for you
* If answering not all may be included

File: sdp.go [hunk 2]
```diff
@@ -435,3 +444,53 @@ func haveApplicationMediaSection(desc *sdp.SessionDescription) bool {
...
+		// populate known remote extmap and handle conflicts.
+		if _, ok := remoteExtMaps[mediaType]; !ok {
+			remoteExtMaps[mediaType] = map[int]sdp.ExtMap{}
+		}
+		for _, attr := range media.Attributes {
+			if attr.Key != "extmap" {
+				continue
+			}
+			em := &sdp.ExtMap{}
+			if err := em.Unmarshal("extmap:" + attr.Value); err != nil {
```
Comment (Sean-Der) [inline sdp.go:482]: I would make this fatal

File: sdp.go [hunk 3]
```diff
@@ -435,3 +444,53 @@ func haveApplicationMediaSection(desc *sdp.SessionDescription) bool {
 
 	return false
 }
+
+func answerExtMaps(remoteExtMaps map[string]map[int]sdp.ExtMap, localMaps map[string][]sdp.ExtMap) map[string][]sdp.ExtMap {
```
Comment (Sean-Der) [inline sdp.go:448]: LGTM! Also make sure to handle IDs here, looks like right now you are just checking URLs?

File: sdp.go [hunk 4]
```diff
@@ -244,6 +244,15 @@ func addTransceiverSDP(d *sdp.SessionDescription, isPlanB bool, mediaEngine *Med
 		return false, nil
 	}
 
+	// Add extmaps
+	fmt.Printf("Type %v", t.kind.String())
```
Comment (Sean-Der) [inline sdp.go:248]: Remove

File: peerconnection_test.go [hunk 5]
```diff
@@ -245,7 +248,7 @@ func TestPeerConnection_GetConfiguration(t *testing.T) {
 	assert.Equal(t, expected.BundlePolicy, actual.BundlePolicy)
 	assert.Equal(t, expected.RTCPMuxPolicy, actual.RTCPMuxPolicy)
 	// TODO(albrow): Uncomment this after #513 is fixed.
-	// See: https://github.com/pion/webrtc/issues/513.
+	// See: https://github.com/pion/webrtc/v2/issues/513.
```
Comment (Sean-Der) [inline peerconnection_test.go:251]: Undo

Comment (jbrady42) [reply peerconnection_test.go:251]: Bad vscode

File: sdp.go (hunk 3 above)
Comment (jbrady42) [reply sdp.go:462]: Exts are added directly from the remote side here so ID will be correct.

**Comment by Sean-Der** [comment]
Hey @jbrady42 

Consider this accepted :) I need to rebase and deal with those test fails myself, but ran out of time tonight :(

thanks!

**Comment by masterada** [comment]
In this form it does not solve https://github.com/pion/webrtc/issues/1255, because the `media.WithTransportCCExtMap()` call is still there. I'm not sure if this mr was meant to solve the linked issue.

**Comment by Sean-Der** [comment]
@masterada :+1: I will make sure to drop that!

Extensions should ONLY be set by users explicitly requesting them. I think `WithTransportCCExtMap` was just a temporary change to unblock things.

File: sdp.go [hunk 6]
```diff
@@ -303,6 +320,14 @@ func populateSDP(d *sdp.SessionDescription, isPlanB bool, isICELite bool, mediaE
 		// RFC 5245 S15.3
 		d = d.WithValueAttribute(sdp.AttrKeyICELite, sdp.AttrKeyICELite)
 	}
+
+	// Add global exts
+	if maps, ok := extMaps["global"]; ok {
```
Comment (pongraczgabor87) [inline sdp.go:325]:
```suggestion
	if maps, ok := extMaps[SDPSectionGlobal]; ok {
```

File: sdp.go [hunk 7]
```diff
@@ -435,3 +460,74 @@ func haveApplicationMediaSection(desc *sdp.SessionDescription) bool {
...
+			if err := maybeAddExt(attr, mediaType); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// Add global exts
+	mediaType := SDPSectionGlobal
+	for _, attr := range session.Attributes {
+		if err := maybeAddExt(attr, mediaType); err != nil {
```
Comment (pongraczgabor87) [inline sdp.go:528]:
```suggestion
	for _, attr := range session.Attributes {
		if err := maybeAddExt(attr, SDPSectionGlobal); err != nil {
```

File: sdp.go [hunk 8]
```diff
@@ -435,3 +450,59 @@ func haveApplicationMediaSection(desc *sdp.SessionDescription) bool {
 
 	return false
 }
+
+func matchedAnswerExt(descriptions []*sdp.MediaDescription, localMaps map[string][]sdp.ExtMap) (map[string][]sdp.ExtMap, error) {
```
[the remark this answers is not in the document]
Comment (jbrady42) [reply sdp.go]: Good catch

**Pull request merged**